tests/integration/test_parallel_replicas_custom_key/test.py

```diff
@@ -22,46 +25,106 @@ def start_cluster():
 
 def create_tables(cluster):
     n1 = nodes[0]
-    n1.query("DROP TABLE IF EXISTS dist_table")
-    n1.query(f"DROP TABLE IF EXISTS test_table ON CLUSTER {cluster}")
+    n1.query("DROP TABLE IF EXISTS dist_table SYNC")
+    n1.query(f"DROP TABLE IF EXISTS test_table ON CLUSTER {cluster} SYNC")
 
+
+def insert_data(table_name, row_num, all_nodes=False):
+    query = (
+        f"INSERT INTO {table_name} SELECT number % 4, number FROM numbers({row_num})"
+    )
+
+    if all_nodes:
+        for n in nodes:
+            n.query(query)
+    else:
+        n1 = nodes[0]
+        n1.query(query)
+
+
+@pytest.mark.parametrize("custom_key", ["sipHash64(key)", "key"])
+@pytest.mark.parametrize("filter_type", ["default", "range"])
+@pytest.mark.parametrize(
+    "cluster",
+    ["test_multiple_shards_multiple_replicas", "test_single_shard_multiple_replicas"],
+)
+def test_parallel_replicas_custom_key_distributed(
+    start_cluster, cluster, custom_key, filter_type
+):
+    for node in nodes:
+        node.rotate_logs()
+
+    row_num = 1000
+
+    n1 = nodes[0]
+    n1.query(f"DROP TABLE IF EXISTS dist_table ON CLUSTER {cluster} SYNC")
+    n1.query(f"DROP TABLE IF EXISTS test_table_for_dist ON CLUSTER {cluster} SYNC")
     n1.query(
-        f"CREATE TABLE test_table ON CLUSTER {cluster} (key UInt32, value String) Engine=MergeTree ORDER BY (key, sipHash64(value))"
+        f"CREATE TABLE test_table_for_dist ON CLUSTER {cluster} (key UInt32, value String) Engine=MergeTree ORDER BY (key, sipHash64(value))"
     )
+
     n1.query(
         f"""
-            CREATE TABLE dist_table AS test_table
+            CREATE TABLE dist_table AS test_table_for_dist
             Engine=Distributed(
                 {cluster},
                 currentDatabase(),
-                test_table,
+                test_table_for_dist,
                 rand()
             )
             """
     )
 
+    insert_data("dist_table", row_num)
+
+    n1.query("SYSTEM FLUSH DISTRIBUTED dist_table")
+
+    expected_result = ""
+    for i in range(4):
+        expected_result += f"{i}\t250\n"
 
-def insert_data(cluster, row_num):
-    create_tables(cluster)
     n1 = nodes[0]
-    n1.query(
-        f"INSERT INTO dist_table SELECT number % 4, number FROM numbers({row_num})"
+    assert (
+        n1.query(
+            "SELECT key, count() FROM dist_table GROUP BY key ORDER BY key",
+            settings={
+                "max_parallel_replicas": 4,
+                "parallel_replicas_custom_key": custom_key,
+                "parallel_replicas_custom_key_filter_type": filter_type,
+                "prefer_localhost_replica": 0,
```

Comment:
It seems we disable `prefer_localhost_replica` during planning - is it necessary here?

Reply:
we want to test the case when custom key is used with offset mode (when multiple shards defined) so we have to avoid generating local plans manually.
`prefer_localhost_replica` is disabled manually only for parallel replicas with custom key if there is a single shard with multiple replicas because it generates a single plan.
In the offset case, we can have shards that don't have local replicas so a plan utilizing multiple replicas will be generated even with `prefer_localhost_replica = 1`